File: ruby/spec/gilded_rose_spec.rb
require 'spec_helper'

# NOTE uncomment one of the lines to run tests for proper GildedRose class
#require File.join(File.dirname(__FILE__) + '/..', 'gilded_rose')
#require File.join(File.dirname(__FILE__) + '/..', 'gilded_rose_refactored_1')
require File.join(File.dirname(__FILE__) + '/..', 'gilded_rose_refactored_2')

describe GildedRose do

  describe '#update_quality' do
    # ######################################################################
    # Shared examples
    # ######################################################################
    shared_examples 'default item sell in' do |item_name|
      it 'lowers sell in value by 1 at the end of the day' do
        item = Item.new(item_name, sell_in=1, quality=0)
        items = [item]
        gilded_rose = described_class.new(items)
        gilded_rose.update_quality

        expect(item.sell_in).to eq 0
      end

      it 'lowers sell in value by N after N days' do
        n = 10
        item = Item.new(item_name, sell_in=n, quality=0)
        items = [item]
        gilded_rose = described_class.new(items)

        n.times do |i|
          gilded_rose.update_quality
          expect(item.sell_in).to eq (n - (i + 1))
        end

        expect(item.sell_in).to eq 0
      end

      it 'sell in value can be negative' do
        item = Item.new(item_name, sell_in=0, quality=0)
        items = [item]
        gilded_rose = described_class.new(items)
        gilded_rose.update_quality

        expect(item.sell_in).to eq -1
      end
    end

    shared_examples 'quality value' do |item_name|
      it 'quality value is never negative' do
        item = Item.new(item_name, sell_in=0, quality=0)
        items = [item]
        gilded_rose = described_class.new(items)
        gilded_rose.update_quality

        expect(item.quality).to eq 0
      end

      it 'quality value is never more than 50' do
        item = Item.new(item_name, sell_in=20, quality=50)
        items = [item]
        gilded_rose = described_class.new(items)
        gilded_rose.update_quality

        expect(item.quality).to be <= 50
      end
    end

    # ######################################################################
    # Default item
    # ######################################################################
    context 'item name' do
      it 'does not change the name' do
        item = Item.new('foo', sell_in=0, quality=0)
        items = [item]
        gilded_rose = described_class.new(items)
        gilded_rose.update_quality

        expect(item.name).to eq 'foo'
      end
    end

    it_behaves_like 'default item sell in', item_name='foo'

    context 'item quality' do
      it_behaves_like 'quality value', item_name='foo'

      context 'when sell in date not passed yet' do
        it 'lowers quality value by 1 at the end of the day' do
          item = Item.new('foo', sell_in=1, quality=1)
          items = [item]
          gilded_rose = described_class.new(items)
          gilded_rose.update_quality

          expect(item.quality).to eq 0
        end

        it 'lowers quality value by N after N days' do
          n = 10
          item = Item.new('foo', sell_in=n, quality=n)
          items = [item]
          gilded_rose = described_class.new(items)

          n.times do |i|
            gilded_rose.update_quality
            expect(item.quality).to eq (n - (i + 1))
          end

          expect(item.quality).to eq 0
        end
      end

      context 'when sell in date has passed' do
        it 'lowers quality value by 2 at the end of the day' do
          item = Item.new('foo', sell_in=0, quality=4)
          items = [item]
          gilded_rose = described_class.new(items)
          gilded_rose.update_quality

          expect(item.quality).to eq 2
        end

        it 'lowers quality value twice as fast after N days' do
          n = 5
          quality = 15
          item = Item.new('foo', sell_in=0, quality=quality)
          items = [item]
          gilded_rose = described_class.new(items)

          n.times do |i|
            gilded_rose.update_quality
            expect(item.quality).to eq (quality - (2 * (i+1)))
          end

          expect(item.quality).to eq (quality - (2 * n))
        end
      end
    end

    # ######################################################################
    # Item: Aged Brie
    # ######################################################################
    context 'when item is Aged Brie' do
      it_behaves_like 'default item sell in', item_name='Aged Brie'

      context 'item quality' do
        context 'when sell in date not passed yet' do
          it 'increases by 1 the older it gets' do
            n = 5
            item = Item.new('Aged Brie', sell_in=n, quality=0)
            items = [item]
            gilded_rose = described_class.new(items)

            n.times do |i|
              gilded_rose.update_quality
              expect(item.quality).to eq (i + 1)
            end

            expect(item.quality).to eq n
          end
        end

        context 'when sell in date has passed' do
          it 'increases twice as fast the older it gets' do
            n = 5
            item = Item.new('Aged Brie', sell_in=0, quality=0)
            items = [item]
            gilded_rose = described_class.new(items)

            n.times do |i|
              gilded_rose.update_quality
              expect(item.quality).to eq (2 * (i + 1))
            end

            expect(item.quality).to eq (2 * n)
          end
        end

        it 'is never more than 50' do
          n = 2
          item = Item.new('Aged Brie', sell_in=n, quality=49)
          items = [item]
          gilded_rose = described_class.new(items)

          n.times do
            gilded_rose.update_quality
          end

          expect(item.quality).to eq 50
        end
      end
    end

    # ######################################################################
    # Sulfuras, Hand of Ragnaros
    # ######################################################################
    context 'when item is Sulfuras, Hand of Ragnaros' do
      context 'item sell in' do
        it 'does not change the sell in' do
          item = Item.new('Sulfuras, Hand of Ragnaros', sell_in=0, quality=0)
          items = [item]
          gilded_rose = described_class.new(items)
          gilded_rose.update_quality

          expect(item.sell_in).to eql 0
        end
      end

      context 'item quality' do
        it_behaves_like 'quality value', item_name='Sulfuras, Hand of Ragnaros'

        it 'does not change the quality' do
          item = Item.new('Sulfuras, Hand of Ragnaros', sell_in=0, quality=0)
          items = [item]
          gilded_rose = described_class.new(items)
          gilded_rose.update_quality

          expect(item.quality).to eql 0
        end
      end
    end

    # ######################################################################
    # Backstage passes to a TAFKAL80ETC concert
    # ######################################################################
    context 'when item is Backstage passes to a TAFKAL80ETC concert' do
      it_behaves_like 'default item sell in', item_name='Backstage passes to a TAFKAL80ETC concert'

      context 'item quality' do
        it_behaves_like 'quality value', item_name='Backstage passes to a TAFKAL80ETC concert'

        context 'when sell in date not passed yet' do
          context 'when sell in above 10 days' do
            it 'increases by 1 the older it gets' do
              n = 5
              quality = 1
              item = Item.new('Backstage passes to a TAFKAL80ETC concert', sell_in=15, quality=quality)
              items = [item]
              gilded_rose = described_class.new(items)

              n.times do |i|
                gilded_rose.update_quality
                expect(item.quality).to eq (quality + i + 1)
              end

              expect(item.quality).to eq (quality + n)
            end
          end

          context 'when sell in 10 days or less and above 5 days' do
            it 'increases by 2 the older it gets' do
              n = 5
              quality = 1
              item = Item.new('Backstage passes to a TAFKAL80ETC concert', sell_in=10, quality=quality)
              items = [item]
              gilded_rose = described_class.new(items)

              n.times do |i|
                gilded_rose.update_quality
                expect(item.quality).to eq (quality + 2 * (i + 1))
              end

              expect(item.quality).to eq (quality + (2 * n))
            end
          end

          context 'when sell in 5 days or less' do
            it 'increases by 3 the older it gets' do
              n = 5
              quality = 1
              item = Item.new('Backstage passes to a TAFKAL80ETC concert', sell_in=5, quality=quality)
              items = [item]
              gilded_rose = described_class.new(items)

              n.times do |i|
                gilded_rose.update_quality
                expect(item.quality).to eq (quality + 3 * (i + 1))
              end

              expect(item.quality).to eq (quality + (3 * n))
            end
          end
        end

        context 'when sell in date has passed' do
          it 'drops to 0 after the concert' do
            item = Item.new('Backstage passes to a TAFKAL80ETC concert', sell_in=0, quality=5)
            items = [item]
            gilded_rose = described_class.new(items)
            gilded_rose.update_quality

            expect(item.quality).to eq 0
          end
        end
      end
    end

    # ######################################################################
    # Conjured Mana Cake
    # ######################################################################
    #
    # This is a new feature not yet implemented in GildedRose
    # use :skip tag to skip those tests
    # use :xskip or remove tag to run those tests
    context 'when item is Conjured Mana Cake', :xskip do
      it_behaves_like 'default item sell in', 'Conjured Mana Cake'

      context 'item quality' do
        it_behaves_like 'quality value', item_name='Conjured Mana Cake'

        it 'lowers quality value by 2 at the end of the day' do
          item = Item.new('Conjured Mana Cake', sell_in=1, quality=2)
          items = [item]
          gilded_rose = described_class.new(items)
          gilded_rose.update_quality

          expect(item.quality).to eq 0
        end

        it 'lowers quality value twice as fast after N days' do
          n = 5
          quality = 15
          item = Item.new('Conjured Mana Cake', sell_in=1, quality=quality)
          items = [item]
          gilded_rose = described_class.new(items)

          n.times do |i|
            gilded_rose.update_quality
            expect(item.quality).to eq (quality - (2 * (i+1)))
          end

          expect(item.quality).to eq (quality - (2 * n))
        end
      end
    end

  end

end

File: js/spec/gilded_rose_spec.js
describe("Gilded Rose", function() {
  var expectNameNotChanged = (name) => {
    it("does not change the name", () => {
      var items = [ new Item(name, 0, 0) ];
      update_quality(items);
      expect(items[0].name).toEqual(name);
    });
  }

  var expectSellInBehaviour = (itemName) => {
    it("lowers sell in value by 1 at the end of the day", () => {
      var items = [ new Item(itemName, 1, 0) ];
      update_quality(items);
      expect(items[0].sell_in).toEqual(0);
    });

    it("lowers sell in value by N after N days", () => {
      var n = 10
      var items = [ new Item(itemName, n, 0) ];

      for (n; n > 0; n--) {
        update_quality(items);
        expect(items[0].sell_in).toEqual(n-1);
      }
      expect(items[0].sell_in).toEqual(0);
    });

    it("sell in value can be negative", () => {
      var items = [ new Item(itemName, 0, 0) ];
      update_quality(items);
      expect(items[0].sell_in).toEqual(-1);
    });
  }

  var expectQualityRange = (itemName) => {
    it('quality value is never negative', () => {
      var items = [ new Item(itemName, 0, 0) ];
      update_quality(items);
      expect(items[0].quality).toBeGreaterThanOrEqual(0);
    });

    it('quality value is never more than 50', () => {
      var items = [ new Item(itemName, 20, 50) ];
      update_quality(items);
      expect(items[0].quality).toBeLessThanOrEqual(50);
    });
  }

  describe('Default item "foo"', () => {
    var name = 'foo';
    expectNameNotChanged(name);
    expectSellInBehaviour(name);

    describe('item quality', () => {
      expectQualityRange(name);

      describe('when sell in date not passed yet', () => {
        it('lowers quality value by 1 at the end of the day', () => {
          var items = [ new Item(name, 1, 1) ];
          update_quality(items);
          expect(items[0].quality).toEqual(0);
        });

        it('lowers quality value by N after N days', () => {
          var n = 10;
          var items = [ new Item(name, n, n) ];

          for (n; n > 0; n--) {
            update_quality(items);
            expect(items[0].quality).toEqual(n-1);
          }
          expect(items[0].quality).toEqual(0);
        });
      });

      describe('when sell in date has passed', () => {
        it('lowers quality value by 2 at the end of the day', () => {
          var items = [ new Item(name, 0, 4) ];
          update_quality(items);
          expect(items[0].quality).toEqual(2);
        });

        it('lowers quality value twice as fast after N days', () => {
          var n = 5;
          var quality = 15;
          var items = [ new Item(name, 0, quality) ];

          for (var i = 1; i <= n; i++) {
            update_quality(items);
            expect(items[0].quality).toEqual(quality - 2*i);
          }
          expect(items[0].quality).toEqual(quality - 2*n);
        });
      });
    });
  });

  describe('Aged Brie', () => {
    var name = 'Aged Brie';
    expectNameNotChanged(name);
    expectSellInBehaviour(name);

    describe('item quality', () => {
      expectQualityRange(name);

      describe('when sell in date not passed yet', () => {
        it('increases by 1 the older it gets', () => {
          var n = 5;
          var items = [ new Item(name, n, 0) ];

          for (var i = 1; i <= n; i++) {
            update_quality(items);
            expect(items[0].quality).toEqual(i);
          }
          expect(items[0].quality).toEqual(n);
        });
      });

      describe('when sell in date has passed', () => {
        it('increases twice as fast the older it gets', () => {
          var n = 5;
          var items = [ new Item(name, 0, 0) ];

          for (var i = 1; i <= n; i++) {
            update_quality(items);
            expect(items[0].quality).toEqual(2*i);
          }
          expect(items[0].quality).toEqual(2*n);
        });
      });

      it('is never more than 50', () => {
        var items = [ new Item(name, 0, 49) ];
        update_quality(items);
        expect(items[0].quality).toEqual(50);
      });
    });
  });

  describe('Sulfuras, Hand of Ragnaros', () => {
    var name = 'Sulfuras, Hand of Ragnaros';
    expectNameNotChanged(name);

    it('does not change the sell in', () => {
      var items = [ new Item(name, 1, 0) ];
      update_quality(items);
      expect(items[0].sell_in).toEqual(1);
    });

    describe('item quality', () => {
      expectQualityRange(name);

      describe('when sell in date not passed yet', () => {
        it('does not change the quality', () => {
          var items = [ new Item(name, 3, 0) ];
          update_quality(items);
          expect(items[0].quality).toEqual(0);
        });
      });

      describe('when sell in date has passed', () => {
          var items = [ new Item(name, -1, 0) ];
          update_quality(items);
          expect(items[0].quality).toEqual(0);
      });
    });
  });

  describe('Backstage passes to a TAFKAL80ETC concert', () => {
    var name = 'Backstage passes to a TAFKAL80ETC concert';
    expectNameNotChanged(name);
    expectSellInBehaviour(name);

    describe('item quality', () => {
      expectQualityRange(name);

      describe('when sell in date not passed yet', () => {
        describe('when sell in above 10 days', () => {
          it('increases by 1 the older it gets', () => {
            var n = 5;
            var items = [ new Item(name, 20, 0) ];

            for (var i = 1; i <= n; i++) {
              update_quality(items);
              expect(items[0].quality).toEqual(i);
            }
            expect(items[0].quality).toEqual(n);
          });;

          it('increases to 50 when sell_in above 10 and quality is 49', () => {
            var items = [ new Item(name, 15, 49) ];
            update_quality(items);
            expect(items[0].quality).toEqual(50);
          });

          it('increases to 50 instead of 51 when sell_in at least 5 and quality is 49', () => {
            var items = [ new Item(name, 5, 49) ];
            update_quality(items);
            expect(items[0].quality).toEqual(50);
          });
        });

        describe('when sell in 10 days or less and above 5 days', () => {
          it('increases by 2 the older it gets', () => {
            var n = 5;
            var quality = 1
            var items = [ new Item(name, 10, quality) ];

            for (var i = 1; i <= n; i++) {
              update_quality(items);
              expect(items[0].quality).toEqual(quality + 2*i);
            }
            expect(items[0].quality).toEqual(quality + 2*n);
          });

          it('increases to 50 instead of 51 when sell_in at least 1 and quality is 49', () => {
            var items = [ new Item(name, 10, 49) ];
            update_quality(items);
            expect(items[0].quality).toEqual(50)
          });
        });

        describe('when sell in 5 days or less', () => {
          it('increases by 3 the older it gets', () => {
            var n = 5;
            var quality = 1
            var items = [ new Item(name, 5, quality) ];

            for (var i = 1; i <= n; i++) {
              update_quality(items);
              expect(items[0].quality).toEqual(quality + 3*i);
            }
            expect(items[0].quality).toEqual(quality + 3*n);
          });

          it('increases to 50 instead of 52 when sell_in at least 1 and quality is 49', () => {
            var items = [ new Item(name, 5, 49) ];
            update_quality(items);
            expect(items[0].quality).toEqual(50)
          });
        });
      });

      describe('when sell in date has passed', () => {
        it('drops to 0 after the concert', () => {
          var items = [ new Item(name, 0, 5) ];
          update_quality(items);
          expect(items[0].quality).toEqual(0)
        });
      });
    });
  });

  describe('Conjured Mana Cake', () => {
    var name = 'Conjured Mana Cake';
    expectNameNotChanged(name);
    expectSellInBehaviour(name);

    describe('item quality', () => {
      expectQualityRange(name);

      //  This is a new feature not yet implemented in gilded rose
      //  use xit() to skip those tests
      //  change to it() to run those tests

      xit('lowers quality value by 2 at the end of the day', () => {
        var items = [ new Item(name, 1, 5) ];
        update_quality(items);
        expect(items[0].quality).toEqual(3)
      });

      xit('lowers quality value twice as fast after N days', () => {
        var quality = 15;
        var items = [ new Item(name, 1, quality) ];

        update_quality(items);
        expect(items[0].quality).toEqual(13)

        update_quality(items);
        expect(items[0].quality).toEqual(9)

        update_quality(items);
        expect(items[0].quality).toEqual(5)

        update_quality(items);
        expect(items[0].quality).toEqual(1)

        update_quality(items);
        expect(items[0].quality).toEqual(0)
      });
    });
  });
});
